### ISSUE TITLE
Saros/I does not end the running session when the shared project is closed

### ISSUE THREAD

**Issue by tobous** [saros-project/saros#582]
Saros/I does not end the running session when the shared project is closed. This leads to issues as it then tries to apply activities to the disposed project/its modules.

This issue can be solved by implementing a project-closed listener which ends the session if the shared project is closed. The only open issue with this approach is how to synchronously end the session or filter remaining activities to guarantee that no activities are applied to disposed resources. This was further discussed in #539 and #568.

**Comment by tobous** [comment]
@srossbach Could you explain to me again why it is a problem to stop the session on the event dispacher thread? Also, is there any documentation on this (or was there some documentation on the old website that did not get migrated)?

Either way, independently of whether I stop the session synchronously on the EDT or not, I still get issues when closing the project while Saros is still processing local activities (i.e. typing something in a shared editor and then immediately closing the editor). It leads to an interrupted exception (see stacktrace below) and I have no idea what to do about it. I tried different (hacky) solutions to avoid this (filtering incoming activities, manually shutting down the local IDE listeners, manually stopping the `ActivityHandler`), but it didn't help. But I am not sure I fully understand the problem, so these approaches might have been inappropriate anyways.

Any ideas on what the issues is and/or to do?
```
ERROR 15:30:23,407 [dpp-activity-dispatcher] (ThreadUtils.java:37) Internal Error:
com.intellij.openapi.progress.ProcessCanceledException: java.lang.InterruptedException
	at com.intellij.util.concurrency.Semaphore.waitFor(Semaphore.java:76)
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:173)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:680)
	at saros.intellij.runtime.IntelliJSynchronizer.exec(IntelliJSynchronizer.java:35)
	at saros.intellij.runtime.IntelliJSynchronizer.syncExec(IntelliJSynchronizer.java:21)
	at saros.session.internal.ActivityHandler.dispatchAndExecuteActivities(ActivityHandler.java:363)
	at saros.session.internal.ActivityHandler.access$200(ActivityHandler.java:33)
	at saros.session.internal.ActivityHandler$1.run(ActivityHandler.java:100)
	at saros.util.ThreadUtils$1.run(ThreadUtils.java:35)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.InterruptedException
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:998)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at com.intellij.util.concurrency.Semaphore.waitForUnsafe(Semaphore.java:81)
	at com.intellij.util.concurrency.Semaphore.waitFor(Semaphore.java:73)
	... 9 more
ERROR 15:30:23,408 [dpp-activity-dispatcher] (ThreadUtils.java:38) Original caller:
StackTrace:
	at saros.util.ThreadUtils.wrapSafe(ThreadUtils.java:29)
	at saros.util.ThreadUtils.runSafeAsync(ThreadUtils.java:58)
	at saros.session.internal.ActivityHandler.start(ActivityHandler.java:242)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at saros.repackaged.picocontainer.lifecycle.StartableLifecycleStrategy.doLifecycleMethod(StartableLifecycleStrategy.java:101)
	at saros.repackaged.picocontainer.lifecycle.StartableLifecycleStrategy.startComponent(StartableLifecycleStrategy.java:96)
	at saros.repackaged.picocontainer.lifecycle.StartableLifecycleStrategy.start(StartableLifecycleStrategy.java:87)
	at saros.repackaged.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.start(AbstractInjectionFactory.java:84)
	at saros.repackaged.picocontainer.behaviors.AbstractBehavior.start(AbstractBehavior.java:169)
	at saros.repackaged.picocontainer.behaviors.Stored$RealComponentLifecycle.start(Stored.java:132)
	at saros.repackaged.picocontainer.behaviors.Stored.start(Stored.java:110)
	at saros.repackaged.picocontainer.DefaultPicoContainer.potentiallyStartAdapter(DefaultPicoContainer.java:1004)
	at saros.repackaged.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:997)
	at saros.repackaged.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:755)
	at saros.session.internal.SarosSession.start(SarosSession.java:614)
	at saros.negotiation.IncomingSessionNegotiation.startSession(IncomingSessionNegotiation.java:307)
	at saros.negotiation.IncomingSessionNegotiation.accept(IncomingSessionNegotiation.java:174)
	at saros.intellij.ui.wizards.JoinSessionWizard$2.run(JoinSessionWizard.java:94)
	at saros.intellij.ui.wizards.Wizard$2.run(Wizard.java:182)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:731)
	at com.intellij.openapi.progress.impl.CoreProgressManager$5.run(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:586)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:86)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:151)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$null$10(ApplicationImpl.java:594)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:311)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
DEBUG 15:30:23,409 [AWT-EventQueue-0] (SarosSessionManager.java:337) terminating all running negotiations
DEBUG 15:30:23,411 [AWT-EventQueue-0] (XStreamExtensionProvider.java:192) Unregistering (resetting) converter of class saros.misc.xstream.UserConverter
DEBUG 15:30:23,412 [AWT-EventQueue-0] (XStreamExtensionProvider.java:152) Unregistering (resetting) converter of class saros.misc.xstream.SPathConverter
DEBUG 15:30:23,415 [AWT-EventQueue-0] (AbstractLocalDocumentModificationHandler.java:88) Stopped listening for document events
DEBUG 15:30:23,416 [AWT-EventQueue-0] (AbstractLocalDocumentModificationHandler.java:88) Stopped listening for document events
DEBUG 15:30:23,419 [BinaryChannel-intellij_test_1] (BinaryChannelConnection.java:96) [mode=SOCKS5 (D), id=saros-main-session] intellij_test_1@saros-con.imp.fu-berlin.de/Saros connection closed locally: Socket closed
DEBUG 15:30:23,420 [AWT-EventQueue-0] (DataTransferManager.java:349) closed connection [pool id=saros-main-session:out:intellij_test_1@saros-con.imp.fu-berlin.de/Saros]: [mode=SOCKS5 (D), id=saros-main-session] intellij_test_1@saros-con.imp.fu-berlin.de/Saros
```

**Comment by srossbach** [comment]
It seems the UISynchronizer implementation of IntelliJ works slightly different. If you call Display#syncExec in SWT it simply ignores all thread interruptions. Seems the IntelliJ version does not. In other words, while your are waiting for execution you got interrupted (and the execution is still pending). However the activity dispatch is shutdown. In this case not gracefully because of the exception.

See https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/ActivityHandler.java#L85

The InterruptedException is only expected when fetching activities out of the queue.

On the other hand if you call SarosSession#Stop from the EDT and this exception would not happen there is a chance that a timeout can occur in the ActivityHandler#Stop method (which is only possible if you call Session#Stop  from an activity consumer).

To avoid this exception you can simply use a Poison Pill approach.

See https://dzone.com/articles/producers-and-consumers-part-3

So the stop method logic is like:

add POISON_PILL

wait for thread termination

if (!terminated)
interrupt thread

wait for thread termination

if (!terminated)
 game over - nothing we can do

**Comment by srossbach** [comment]
BTW there was a hint in the AWT implementation which you had removed some time ago.

https://github.com/saros-project/saros/commit/d735e04246ed18f3dd360b47f3a07e0dcfad59b5#diff-daee375bbfb7fe922e8f7e8a2e336542L27

**Comment by tobous** [comment]
Sorry, but I am totally lost on this. Could you please explain it to me in more detail?

Which thread/runnable should exactly be ended through the poison pill approach? The ActivityHandler (i.e. the `dispatcherThreadRunnable`)?

The strategy mentioned in the [linked AWT implementation](https://github.com/saros-project/saros/commit/d735e04246ed18f3dd360b47f3a07e0dcfad59b5#diff-daee375bbfb7fe922e8f7e8a2e336542L27) (if I understand it correctly) does not work as the synchronized execution method is also called from the EDT, which would lead to a deadlock (at least the way I understand it):
1. wrap runnable
2. execute wrapped runnable asynchronously on EDT
3. on thread that called the synchronous execution, wait on lock
4. release lock in wrapped runnable after actual runnable is done on EDT
    - include check whether wrapped runnable is already done before wait call

The if the calling thread in step 3 is already the EDT, the wait call will block the EDT, meaning the asynchronous execution on the EDT that would eventually release the initial thread can never be run/scheduled.

**Comment by tobous** [comment]
@srossbach Could you provide more information on the topic?

**Comment by srossbach** [comment]
The dispatchThreadRunnable. And of course you will get a TimeoutException if you call this while on the EDT that is why you SHOULD NOT stop the session from the EDT.

As you do not have access to https://docs.oracle.com/javase/7/docs/api/java/awt/SecondaryLoop.html in IntelliJ I do not know what your options are to block the EDT while shutting down the session in the background.
